File: packages/igloo-ts/src/widgets/vertical-slider.ts
import { VerticalSlider as WitVerticalSlider } from 'iced:app/vertical-slider@0.1.0';
import type { CallbackId } from 'iced:app/callbacks@0.1.0';
import type { Length } from 'iced:app/length@0.1.0';
import type { Pixels } from 'iced:app/shared@0.1.0';
import { Element, type IntoElement } from '../element.js';
import { pushF32, pushFixed } from '../callbacks.js';

/**
 * Builder for creating vertical Slider widgets.
 * A VerticalSlider is a vertical bar for selecting a value from a range.
 *
 * @example
 * ```typescript
 * const slider = VerticalSlider.new(0, 100, state.volume, (value) => ({
 *   type: 'volumeChanged',
 *   value,
 * })).step(1);
 * ```
 *
 * @typeParam Msg - The message type this slider emits; inferred from `onChange`.
 */
export class VerticalSlider<Msg = never> implements IntoElement<Msg> {
  private raw: WitVerticalSlider;

  private constructor(rangeStart: number, rangeEnd: number, value: number, onChange: CallbackId) {
    this.raw = new WitVerticalSlider(rangeStart, rangeEnd, value, onChange);
  }

  /**
   * Create a new VerticalSlider builder.
   * @param onChange - Handler called with the new value while dragging
   */
  static new<const M>(
    rangeStart: number,
    rangeEnd: number,
    value: number,
    onChange: (value: number) => M,
  ): VerticalSlider<M> {
    return new VerticalSlider(rangeStart, rangeEnd, value, pushF32(onChange));
  }

  /** Set the default value (value to reset to on double-click) */
  default(defaultValue: number): this {
    this.raw.default(defaultValue);
    return this;
  }

  /** Set the message to emit when the slider is released */
  onRelease<const M>(msg: () => M): VerticalSlider<Msg | M> {
    this.raw.onRelease(pushFixed(msg()));
    return this as unknown as VerticalSlider<Msg | M>;
  }

  /** Set the width in pixels */
  width(width: Pixels): this {
    this.raw.width(width);
    return this;
  }

  /** Set the height */
  height(height: Length): this {
    this.raw.height(height);
    return this;
  }

  /** Set the step size for normal dragging */
  step(step: number): this {
    this.raw.step(step);
    return this;
  }

  /** Set the step size when holding shift */
  shiftStep(step: number): this {
    this.raw.shiftStep(step);
    return this;
  }

  /** Convert to Element */
  intoElement(): Element<Msg> {
    return new Element(WitVerticalSlider.intoElement(this.raw));
  }
}

File: crates/igloo_guest/src/widgets/vertical_slider.rs
use std::ops::RangeInclusive;

use iced_core::{Length, Pixels};

use crate::Element;
use crate::bindings::iced::app::vertical_slider::VerticalSlider as WitVerticalSlider;

/// An interactive vertical bar for selecting a value from a range.
pub struct VerticalSlider<Message> {
    range: (f32, f32),
    value: f32,
    on_change: Box<dyn Fn(f32) -> Message>,
    on_release: Option<Message>,
    width: Option<Pixels>,
    height: Option<Length>,
    step: Option<f32>,
    shift_step: Option<f32>,
    default: Option<f32>,
}

impl<Message: 'static> VerticalSlider<Message> {
    /// Creates a new [`VerticalSlider`].
    pub fn new(
        range: RangeInclusive<f32>,
        value: f32,
        on_change: impl Fn(f32) -> Message + 'static,
    ) -> Self {
        Self {
            range: range.into_inner(),
            value,
            on_change: Box::new(on_change),
            on_release: None,
            width: None,
            height: None,
            step: None,
            shift_step: None,
            default: None,
        }
    }

    /// Sets the message to produce when the [`VerticalSlider`] is released.
    pub fn on_release(mut self, message: Message) -> Self {
        self.on_release = Some(message);
        self
    }

    /// Sets the width of the [`VerticalSlider`].
    pub fn width(mut self, width: impl Into<Pixels>) -> Self {
        self.width = Some(width.into());
        self
    }

    /// Sets the height of the [`VerticalSlider`].
    pub fn height(mut self, height: impl Into<Length>) -> Self {
        self.height = Some(height.into());
        self
    }

    /// Sets the step size of the [`VerticalSlider`].
    pub fn step(mut self, step: f32) -> Self {
        self.step = Some(step);
        self
    }

    /// Sets the step size of the [`VerticalSlider`] when the shift key is pressed.
    pub fn shift_step(mut self, step: f32) -> Self {
        self.shift_step = Some(step);
        self
    }

    /// Sets the default value of the [`VerticalSlider`].
    pub fn default(mut self, value: f32) -> Self {
        self.default = Some(value);
        self
    }
}

impl<Message: 'static> From<VerticalSlider<Message>> for Element<Message> {
    fn from(slider: VerticalSlider<Message>) -> Self {
        Element::new(move |realize| {
            let mapper = realize.f32_mapper(slider.on_change);
            let (start, end) = slider.range;
            let raw = WitVerticalSlider::new(start, end, slider.value, mapper);
            if let Some(msg) = slider.on_release {
                raw.on_release(realize.fixed(msg));
            }
            if let Some(width) = slider.width {
                raw.width(width.0);
            }
            if let Some(height) = slider.height {
                raw.height(height.into());
            }
            if let Some(step) = slider.step {
                raw.step(step);
            }
            if let Some(step) = slider.shift_step {
                raw.shift_step(step);
            }
            if let Some(value) = slider.default {
                raw.default(value);
            }
            WitVerticalSlider::into_element(raw)
        })
    }
}
